### ISSUE TITLE
Suggestion: Try better low-rank approximation layer type

### ISSUE THREAD

**Issue by catid** [r-three/phatgoose#1]
https://github.com/r-three/phatgoose/blob/2c841d70724192401598ed9d2c94d088c9cd3377/src/models/custom_modules/lora/lora_linear.py#L8

This probably will be improved by using this one instead:

```
def round_up_sqrt(m):
    sm = int(m ** 0.5 + 0.5)
    while sm*sm > m:
        sm -= 1
    while sm*sm < m:
        sm += 1
    return sm

# Stacked Kronecker-product Layers https://openreview.net/pdf?id=ZjGr1tMVbjw
# Uses 2r*sqrt(nm) parameters instead of nm.
# For for n=512 x m=2048, r must be 256 or less to make it worthwhile.
class SKLinear(nn.Module):
    def __init__(self, n, m, scale=0.5):
        super().__init__()
        self.n = n
        self.m = m
        self.sn = round_up_sqrt(n)
        self.np = self.sn * self.sn # n rounded up to next square
        self.sm = round_up_sqrt(m)
        self.mp = self.sm * self.sm # m rounded up to next square
        k = self.sn * self.sm

        r = int(scale * (n * m) / 2.0 / k + 0.5)

        print(f"Using SKLinear: n={n} m={m} r={r} round_sqrt(n)={self.sn} round_sqrt(m)={self.sm} n'={self.np} m'={self.mp} k={k} reduction={(2 * r * k) * 100.0 / (n * m)}%")

        # Initialize A and B using Kaiming initialization
        self.A = nn.Parameter(torch.empty(k, r))
        nn.init.kaiming_uniform_(self.A, a=math.sqrt(5)) # a is the parameter for the ReLU
        self.B = nn.Parameter(torch.empty(r, k))
        nn.init.kaiming_uniform_(self.B, a=math.sqrt(5))

    def forward(self, x):
        # Validate that the inputs are of the expected sizes
        if x.size(-1) != self.n:
            raise ValueError("Input vector must have size n")

        S = torch.matmul(self.A, self.B).reshape(self.sn, self.sm, self.sn, self.sm).transpose(1, 2).reshape(self.np, self.mp)
        S_truncated = S[:self.n, :self.m]

        return torch.matmul(x, S_truncated)
```

**Comment by catid** [comment]
In my testing, for the same number of weights as the naive version, you get much better accuracy scores.
